lib/meilisearch/tenant_token.rb

```diff
@@ -0,0 +1,72 @@
+# frozen_string_literal: true
+
+require 'openssl/hmac'
+
+module MeiliSearch
+  module TenantToken
+    HEADER = {
+      typ: 'JWT',
+      alg: 'HS256'
+    }.freeze
+
+    def generate_tenant_token(search_rules, api_key: nil, expires_at: nil)
+      signature = retrieve_valid_key!(api_key, @api_key)
+      expiration = validate_expires_at!(expires_at)
+      rules = validate_search_rules!(search_rules)
+      unsigned_data = build_payload(expiration, rules, signature)
+
+      combine(unsigned_data, to_base64(sign_data(signature, unsigned_data)))
+    end
+
+    private
+
+    def build_payload(expiration, rules, signature)
+      payload = {
+        searchRules: rules,
+        apiKeyPrefix: signature[0..7],
+        exp: expiration
+      }.compact
+
+      combine(encode(HEADER), encode(payload))
+    end
+
+    def validate_expires_at!(expires_at)
+      return unless expires_at
+      return expires_at.to_i if expires_at > Time.now
+
+      raise ExpiredSignature
+    rescue StandardError
+      raise ExpiredSignature
+    end
+
+    def validate_search_rules!(data)
+      return data if data
+
+      raise InvalidSearchRules
+    end
+
+    def retrieve_valid_key!(*keys)
+      key = keys.compact.find { |k| !k.empty? }
```

Comment:
Are you checking the given key is an already existing key in the search engine?
the list of the keys is only retrievable if you pass the master key. What if the users want to use `generate_tenant_token` but don't have the primary key?

Reply:
> Are you checking the given key is an already existing key in the search engine?
Actually not, the `generate_tenant_token` method does not make an external request.

What I tried to achieve with this method, is to enable an optional behavior:

If the user wants to generate a token with a particular key, they can: `client.generate_tenant_token(rules, api_key: new_key)`.
Otherwise, we will use the "default" key `client.generate_tenant_token(rules)`, which means, we will use the `@api_key` defined when the user instantiated the client: `client = MeiliSearch::Client.new(url, "masterKey")` (we will use the "masterKey") then.

Reply:
Sorry we're talking about this earlier and I completely forgot that it was only possible to check the other keys with the master key.